### DISCUSSION TITLE
Checkpointing in terms of our language

### DISCUSSION THREAD

**Discussion by tomjaguarpaw** [Mikolaj/mostly-harmless#20] [Ideas]
# Checkpointing

Checkpointing is a well-known technique to reduce peak memory usage of automatically differentiated code at the cost of additional run time. This note explains how we can augment our [Fig. 9 transformation](https://www.microsoft.com/en-us/research/uploads/prod/2021/08/higher-order-ad.pdf) with checkpointing.

## The problem that checkpointing solves

Consider the program

``` haskell
f :: R -> R
f x = let y = expr1 in e2
```

(where `e1 :: R` and `e2 :: R` may depend on `x`). Our translation (Fig 9 of [our paper](https://www.microsoft.com/en-us/research/uploads/prod/2021/08/higher-order-ad.pdf)) transforms it into

``` haskell
D{f} :: (R, Delta) -> M (R, Delta)
D{f} x = do { y <- D{expr1}; D{e2} }

-- where
-- D{expr1} :: M (R, Delta)
-- D{e2} :: M (R, Delta)
```

Evaluating `D{f}` applied to an argument proceeds by first evaluating `D{e1}`, which produces a `DeltaBinds` (i.e. a `[(DeltaId, Delta)]`) inside the monad `M`. Second, `D{e2}` is evaluated, adding some more `DeltaBinds` on top. If the `DeltaBinds` produced this way is a long list this could cause a problem: high peak memory usage (the peak occurs at the end of the evaluation of `D{f}`).

## An example of checkpointing

How can we reduce peak memory usage?  We can do so by using a technique called “checkpointing”. Checkpointing is an automatic transformation that is part of many AD implementations. Before describing how to add checkpointing to our Fig. 9 algorithm (in "General form of checkpointing" below) lets see how we can write by hand a differentiated version of `f` with reduced peak memory usage.

Call the `DeltaBinds` produced by evaluating `D{e1}` `d1` and the `DeltaBinds` produced by evaluating `D{e2}` `d2`. Evaluating `D{f}` uses memory proportional to the size of `d1` *plus* the size of `d2`. We're going to write a new function `D{f}_adjusted` which calculates the same result as `D{f}` but uses memory proportional to the *maximum* of the size of `d1` and the size of `d2`. We will pay for this peak memory saving with increased run time.

We have to change our setup a bit. Rather than

``` haskell
type DeltaBinds = [(DeltaId, Delta)]
```

instead define

``` haskell
type DeltaBinds = [Either (DeltaId, Delta) (M ())]
```

That is, an element of `DeltaBinds` is either a binding known now, or an action in `M` which is a “delayed” or “suspended” computation which will produce some `DeltaBinds` later.

Then instead of our original program transformation, we can produce

``` haskell
f :: R -> R
f x = let y = expr1 in e2

D{f} :: (R, Delta) -> M (R, Delta)
D{f} x = do { y <- D{expr1}; D{e2} }

-- where
-- D{expr1} :: M (R, Delta)
-- D{e2} :: M (R, Delta)

D{f}_alternative :: (R, Delta) -> M (R, Delta)
D{f}_alternative x = do
    y <- do
        yd <- fresh
        delay (do
          (_, yd_inner) <- D{expr1}
          appendDeltaLet yd yd_inner)
        return ((\(x, _) -> expr1) x, Var yd)
    D{e2}
```

That is, instead of running `e1` and generating its `DeltaBinds` now we “delay” generating its `DeltaBinds` until later (see `runDelta'`). We also generate `yd :: DeltaId` to do the impedance matching between `e2` and the delayed `DeltaBinds`.

Here are the required auxiliary functions

``` haskell
delay :: M () -> M ()
delay m = M (\(deltaId, bs) -> ((), (deltaId, Right m : bs)))

fresh :: M DeltaId
fresh = M (\(delta_id, bs) -> (delta_id, (delta_id + 1, bs)))

appendDeltaLet :: DeltaId -> Delta -> M ()
appendDeltaLet did d =
  M (\(delta_id, bs) -> ((), (delta_id, Left (did, d) : bs)))

-- We can rewrite deltaLet in terms of the above
deltaLet :: Delta -> M DeltaId
deltaLet delta = do
  delta_id <- fresh
  appendDeltaLet delta_id delta
  pure delta_id
```

How do we process the resulting `DeltaBinds`, i.e. how do we do the equivalent of `runDelta` and `eval`? The first thing to note is that there is no way to express the new “delayed component” of `DeltaBinds` in `Delta`. The delayed component corresponds to a sequence of `Let`s with no body, and our `Let` always has a body. But that’s OK: there is no need to keep the `Let` constructor in `Delta` anyway. The only purpose it served was to nest `Let`s in the right hand side of other `Let`s. We never did that, so we can drop it and use `DeltaBinds` itself to express sharing instead.

Once we do that `Delta` looks like:

``` haskell
-- Let constructor removed
data Delta = Zero | Scale R Delta | Add Delta Delta | Var DeltaId
```

Then what `eval` used to do on `Let` is done in `evalLet` instead. We can write `runDeltaEval` which combines our original `runDelta` and `eval`:

``` haskell
type DeltaMap = Map DeltaId R

-- Same as before but without Let
eval :: R -> Delta -> DeltaMap -> DeltaMap
eval _ Zero um = um
eval x (Scale y u) um = eval (x * y) u um
eval x (Add u1 u2) um = eval x u2 (eval x u1 um)
eval x (Var uid) um = addDelta uid x um

-- Same as eval used to do on Let
evalLet :: (DeltaId, Delta) -> DeltaMap -> DeltaMap
evalLet (uid, u) um =
  case lookup uid um of
    Nothing -> um
    Just x -> eval x u (delete uid um)

runDelta' :: DeltaId -> DeltaBinds -> DeltaMap -> DeltaMap
runDelta' deltaId deltaBinds deltaMap =
  case deltaBinds of
    [] -> deltaMap
    Left uid_u : rest -> runDelta' deltaId rest (evalLet uid_u deltaMap)
    -- If we come across a suspended `DeltaBinds`
    -- then we run it.
    Right (M m) : rest ->
      let ((), (deltaIdNext, rest')) = m (deltaId, rest)
      in runDelta' deltaIdNext rest' deltaMap

runDeltaEval :: DeltaId -> M (r, Delta) -> DeltaMap -> DeltaMap
runDeltaEval deltaId (M m) deltaMap =
  let ((_r, delta), (deltaIdNext, binds)) = m (deltaId, [])
  in runDelta' deltaIdNext binds (eval 1 delta deltaMap)
```

## Why did we do this?

Call the `DeltaBinds` produced by `D{e1}` and `D{e2}` `d1` and `d2` respectively. The `D{f}` produced by our original (Fig 9/Fig 10) algorithm has peak memory usage that is proportional to the sum of the size of `d1` and `d2`. The alternative transformed `D{f}_alternative`, by contrast, has peak memory usage proportional to the *maximum* of the size of `d1` and `d2` because it *delays* the computation of `d1`, only producing `d2` initially. `runDelta'` consumes `d2` and only when it is finished does it produce `d1`, proceeding to consume it to produce its final result. If `d1` and `d2` are roughly the same size then we have roughly halved our peak memory usage!

To pay for this reduced peak memory usage `D{f}_alternative` has to run `e1_plugged` *as well as* `D{e1}`. `D{f}`, produced by our original algorithm, would only have run `D{e1}`. So we obtain lower peak memory usage but must pay with increased run time.

## General form of checkpointing

Add to “our language” an expression `checkpoint e :: R` for any `e :: R`. Its Fig 9 rule under D (i.e. curly D backwards arrow above) is

``` haskell
D{checkpoint e} :: M (R, Delta)
D{checkpoint e} = do
    ed <- fresh
    delay (do
      (_, ed_inner) <- e1
      appendDeltaLet ed ed_inner)
    return (e_plugged, Var ed)
```

TODO: define `e_plugged`

We can extend this to the general case of `checkpoint e :: T` for any `e :: T` but requires a bit more fiddling to implement the impedance matching.

### Using checkpointing

The `D{f}_adjusted` in "An example of checkpointing" above is then just `D{f}` where

```haskell
f x = let y = checkpoint e1 in e2
```

## References

* <https://engineering.purdue.edu/~qobi/papers/oms2018.pdf>
* <https://www-sop.inria.fr/tropics/papers/DauvergneHascoet06.pdf>

**Comment by Mikolaj** [comment]
Cool. I will read with pleasure and think how this fits our current library code (I will not start implementing until I consult with you, though). 

Please let me know when you decide it's time to include this (or any other point) into the engineering track agenda --- especially that we don't have the theory meeting for the type being.

**Reply by tomjaguarpaw** [reply]
Ah, I didn't know we don't have the theory track. Checkpointing fits best into the theory track, but I'm happy to discuss it any time we don't have more important things to discuss.

I'm also not sure how much checkpointing is used in practice in ML training. It's a big topic in "old school" AD but I haven't heard of PyTorch or Tensorflow doing checkpointing, for example (maybe I just missed it).

**Reply by Mikolaj** [reply]
> I'm also not sure how much checkpointing is used in practice in ML training.

It would be good to ask @awf about that.

> It's a big topic in "old school" AD but I haven't heard of PyTorch or Tensorflow doing checkpointing, for example (maybe I just missed it).

I have no idea, either. If PyTorch or Tensorflow are a bad at memory management (e.g., due to the mixture of C and Python) and store some unneeded intermediate results too long, checkpointing might help. But in the context of Haskell, it seems the checkpointing is kind of explicit data laziness. We know that laziness in some cases leads to memory leaks, which may offset the memory savings. So, for Haskell, I'd need to see experiments to be convinced it does not make the situation worse or unpredictable (as data laziness does, though the laziness being explicit makes it more predictable, because GHC doesn't get to decide whether to retain things).

I mean, the maximum checkpoint (don't create any Delta) is ~probably~ a real saving, because it refers only to the input data and current parameters and both of these are always retained, so having to keep them is not a leak. This maximum checkpoint would defer the creation of `DeltaId_3` in the screenshot at https://github.com/Mikolaj/mostly-harmless/discussions/19#discussioncomment-2283633. However such a checkpoint is useless, because when you force it, you pay the full memory price.

Any checkpoints for the other `DeltaId` in the screenshot would have to retain some of the intermediate vectors (tensors, in general) computed in the forward run (upper-right side of the screenshot, e.g., the `z2` vector that is normally not retained anywhere). I guess one could demonstrate that for the usual operations the leak from a single checkpoint is at most a tensor of size O(size_of_parameters), while the saving is additionally multiplied by O(runtime). But runtime (e.g., the number of layer) is minuscule compared to the size of parameters, so that's still a questionable gain. [Edit: perhaps a better argument is that the number of layers doesn't grow as fast as the size of parameters when scaling neutral networks to harder tasks. This may not be true for RNN for NLP.]

**Comment by tomjaguarpaw** [comment]
> > I'm also not sure how much checkpointing is used in practice in ML training.
>
> It would be good to ask @awf about that.

Yes, good idea.

Checkpointing [is supported in PyTorch](https://pytorch.org/docs/stable/checkpoint.html) I just don't know if it's used much. In any case checkpointing has a long history in AD and there's a lot of literature about when to use it.
